radiation/TwoStreamMLPEMod.F90

```diff
@@ -0,0 +1,1716 @@
+    do_can: do ican = 1,this%n_lyr
+       do_col: do icol = 1,this%n_col(ican)
+          associate(scelg => this%scelg(ican,icol), &
+                    scelb => this%band(ib)%scelb(ican,icol))
+            write(log_unit,*) '--',ican,icol,'--'
+            write(log_unit,*) 'pft:',scelg%pft
+            write(log_unit,*) 'area: ',scelg%area
+            write(log_unit,*) 'lai,sai: ',scelg%lai,scelg%sai
+            write(log_unit,*) 'Kb: ',scelg%Kb
+            write(log_unit,*) 'Kb leaf: ',scelg%Kb_leaf
+            write(log_unit,*) 'Kd: ',scelg%Kd
+            write(log_unit,*) 'Rb0: ',scelb%Rbeam0
+            write(log_unit,*) 'om: ',scelb%om
+            write(log_unit,*) 'betad: ',scelb%betad
+            write(log_unit,*) 'betab:',scelb%betab
+            write(log_unit,*) 'a: ',scelb%a
+            this%band(ib)%Rbeam_atm = 1.0_r8
+            this%band(ib)%Rdiff_atm = 1.0_r8
+            write(log_unit,*)'RDiff Down @ bottom: ',this%GetRdDn(ican,icol,ib,scelg%lai+scelg%sai)
+            write(log_unit,*)'RDiff Up @ bottom: ',this%GetRdUp(ican,icol,ib,scelg%lai+scelg%sai)
+            write(log_unit,*)'Rbeam @ bottom: ',this%GetRb(ican,icol,ib,scelg%lai+scelg%sai)
+          end associate
+       end do do_col
+    end do do_can
+
+  end subroutine Dump
+
+
+  ! ================================================================================================
+
+  subroutine ParamPrep()
+
+    real(r8) :: avmu   ! average inverse optical depth (see Eq 3.4 CLM50 tech man)
+    integer  :: ft
+    integer  :: nbands
+    integer  :: numpft
+    integer  :: ib
+
+    numpft = ubound(rad_params%om_leaf,2)
+    nbands = ubound(rad_params%om_leaf,1)
+    
+    do ft = 1,numpft
+
+       ! The non-band specific parameters here will be re-derived for each
+       ! band, which is inefficient, however this is an incredibly cheap
+       ! routine to begin with, its only called during initialization, so
+       ! just let it go, dont worry about it.
+
+       if(rad_params%xl(ft)<-0.4_r8 .or. rad_params%xl(ft)>0.6_r8) then
+          write(log_unit,*)"Leaf orientation factors (xl) should be between -0.4 and 0.6"
+          write(log_unit,*)"ft: ",ft,"xl: ",rad_params%xl(ft)
+          call endrun(msg=errMsg(sourcefile, __LINE__))
+       end if
+
+       ! There is a singularity of leaf orientation is exactly 0
+       ! phi1 = 0.5
+       ! phi2 = 0.0
+       ! avmu = 1/0  (1 - 0.5/0 * ln(0.5/0.5) ) but the limit approaches 1
+       ! a value of 0.0001 does not break numerics and generates an avmu of nearly 1
+
+       if( abs(rad_params%xl(ft)) <0.0001) rad_params%xl(ft)=0.0001_r8
+       
+       ! There must be protections on xl to prevent div0 and other weirdness
+       rad_params%phi1(ft) = 0.5_r8 - 0.633_r8*rad_params%xl(ft) - 0.330_r8*rad_params%xl(ft)*rad_params%xl(ft)
+       rad_params%phi2(ft) = 0.877_r8 * (1._r8 - 2._r8*rad_params%phi1(ft)) !0 = horiz leaves, 1 - vert leaves.
+       
+       avmu = (1._r8/rad_params%phi2(ft))* &
```

Comment:
I haven't fully mapped the order of operations for the 2-stream, but would it make more sense to define avmu as part of rad_params? It seems avmu is calculated in multiple locations (at least here and in `ZenithPrep`), and having it in a single location might be better to ensure it is always defined consistently.

Reply:
great idea, will do

Reply:
done